src/rgw/rgw_bucket.cc

```diff
@@ -2129,40 +2132,613 @@ void rgw_data_change_log_entry::decode_json(JSONObj *obj) {
+		 << ": " << ec.message() << dendl;
+    }
+    return ceph::from_error_code(ec);
+  }
+  int list(int index, int max_entries,
+	   std::vector<rgw_data_change_log_entry>& entries,
+	   std::optional<std::string_view> marker,
+	   std::string* out_marker, bool* truncated) override {
+    bs::error_code ec;
+    auto [log_entries, more] = fifos[index]->list(max_entries, marker,
+						  ca::use_blocked[ec]);
+    if (ec) {
+      lderr(cct) << __PRETTY_FUNCTION__
+		 << ": failed to list fifo " << get_oid(index)
+		 << ": " << ec.message() << dendl;
+      return ceph::from_error_code(ec);
+    }
+    for (const auto& entry : log_entries) {
+      rgw_data_change_log_entry log_entry;
+      log_entry.log_id = entry.marker;
+      log_entry.log_timestamp = entry.mtime;
+      auto liter = entry.data.cbegin();
+      try {
+	decode(log_entry.entry, liter);
+      } catch (const buffer::error& err) {
+	lderr(cct) << __PRETTY_FUNCTION__
+		   << ": failed to decode data changes log entry: "
+		   << err.what() << dendl;
+	return -EIO;
+      }
+      entries.push_back(std::move(log_entry));
+    }
+    if (truncated)
+      *truncated = more;
+    if (out_marker && !log_entries.empty()) {
+      *out_marker = log_entries.back().marker;
+    }
+    return 0;
+  }
+  int get_info(int index, RGWDataChangesLogInfo *info) override {
+    auto& fifo = fifos[index];
+    bs::error_code ec;
+    fifo->read_meta(ca::use_blocked[ec]);
+    if (ec) {
+      lderr(cct) << __PRETTY_FUNCTION__
+		 << ": failed to read FIFO metadata " << get_oid(index)
+		 << ": " << ec.message() << dendl;
+      return ceph::from_error_code(ec);
+    }
+    auto m = fifo->meta();
+    auto p = m.head_part_num;
+    if (p < 0) {
+      info->marker = neorados::cls::fifo::marker{}.to_string();
+      info->last_update = ceph::real_clock::zero();
+      return 0;
+    }
+    auto h = fifo->get_part_info(p, ca::use_blocked[ec]);
+    if (ec) {
+      lderr(cct) << __PRETTY_FUNCTION__
+		 << ": failed get part info: " << get_oid(index) << "/" << p
+		 << ": " << ec.message() << dendl;
+    } else {
+      info->marker = neorados::cls::fifo::marker{p, h.max_ofs}.to_string();
+      info->last_update = h.max_time;
+    }
+    return ceph::from_error_code(ec);
+  }
+  int trim(int index, std::string_view marker) override {
+    bs::error_code ec;
+    fifos[index]->trim(marker, ca::use_blocked[ec]);
+    if (ec) {
+      lderr(cct) << __PRETTY_FUNCTION__
+		 << ": failed get part info: " << get_oid(index)
+		 << ": " << ec.message() << dendl;
+    }
+    return ceph::from_error_code(ec);
+  }
+  int trim(int index, std::string_view marker,
+	   librados::AioCompletion* c) override {
+    fifos[index]->trim(marker, c);
+    return 0;
+  }
+  std::string_view max_marker() const override {
+    static const std::string mm =
+      neorados::cls::fifo::marker::max().to_string();
+    return std::string_view(mm);
+  }
+};
+
+class RGWDataChangesLegacyFIFO final : public RGWDataChangesBE {
```

Comment:
are you satisfied with testing here? i really don't want to merge both `src/cls/fifo` and `src/neorados/cls/fifo` to master, because then we have to maintain both and ensure they stay consistent. i'd rather see only `src/neorados/cls/fifo` in master, and `src/cls/fifo` in backports that include an upgrade test to verify compatibility